### ISSUE TITLE
`Using Script` directive

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#23890]
### use case:

I am busy with creating a module that includes a lot of classes and enums. Some of these classes are depended on other classes and some classes are even interdepended. I have put a bounty on the related StackOverflow question: [Is it possible to declare two interdependent classes each in a separate file?](https://stackoverflow.com/questions/68657692/is-it-possible-to-declare-two-interdependent-classes-each-in-a-separate-file) and asked around at a few sites as [`#141` Handling interdependent classes](https://github.com/RamblingCookieMonster/PSDepend/issues/141) but haven't been able to find a satisfying answer.
Using a [ModuleBuilder](https://github.com/PoshCode/ModuleBuilder) doesn't appear help as well: I might put the classes in separate script files but that still leaves me with issues as:
* VSCode errors
* [To force classes to be in a certain order, you require to prefix their file names with numbers, like `01-User.ps1`](https://github.com/PoshCode/ModuleBuilder?tab=readme-ov-file#organizing-your-module)
* No solution for *interdepended* classes
* Rebuild the module every time I just need to test a single function/class

Currently, I have decided (for consistency) to put *all* classes in a single `.ps1` file (independent of their dependance) but meanwhile my classes script, is about 2000 lines of code and losing the oversight.

### Proposed technical implementation details (optional)

Somehow I expect that a kind of `Using Script .\MyScript.ps1` directive might could help but I have really a PowerShell script author here and I have no clue how this exactly would work in comparison to the `Using Module` directive and how this would interact with code editors as `VSCode`.

Meaning, I would expect from a `Using Script` directive that scripts are combined loaded by the engine by adding them to a kind of "using-script-list" prior executing the script and than load them if they where part of a single script file (I presume here a similar behavior as the `Using Module` directive).
* If the `Using Script` refers to another script with a `Using Script` directive that script is taken as well
  * Note that in case of an interdepended class, there will be cycling `Using Script` directive, therefore the recursive check should stop if the script is already in the "using-script-list".
 * If the list is complete, the scripts in the "using-script-list" will be (re)loaded as if all the concerned scripts would have been in a single script file.
   * Where scripts in the "using-script-list" will only be reloaded if they have been modified (using the script's `modified date` property)

The difference with the [usual script (dot-source) inclusion](https://stackoverflow.com/q/6016436/1701026) will be:
* Scripts might refer to each other
  * note that dot-sourcing one script in another and vice versa will cause the PowerShell engine to hang
* Scripts will only be reloaded when they changed (I believe that `Using module` has a similar constrain)
* The debugger is aware and therefore prevents an editor to report errors along with "*Unable to find type [<typename>}*" errors

**Comment by iRon7** [comment]
@rhubarb-geek-nz,

> Interesting situation. Before we go looking for changes to PowerShell I suggest first look at the problem.

I wasn't searching for workaround to the problem here but merrily describing the background. Believe me, I have already spend a lot of time searching and posting questions on websites along with StackOverflow finding a proper way to handle a large set of (inter)dependent *PowerShell* classes and moving things back and forth based on all kind of suggestions.
And several features/articles of the existing modulebuilders suggest that my use case doesn't stand on its own.

> you could define common interfaces in a C# module which is shared by your scripts

Not sure where this is actually heading to, but the biggest problem with this is that I am a **PowerShell**  developer and I don't have enough knowledge of C# to build interfaces of the rather complex PowerShell classes I am current working on.

Anyways, the feature request I am presenting here could be possible way out of this issue. If that is not the case or the implementation is questionable or not doable, I simple accept the fact if the request is rejected and closed as such.

Btw. a additional feature (as apposed to dot-sourcing) of a `Using Script` directive could be that it only includes specific statements as `function`, `class`, `enum` and `filter` which might than allow for a securer way to load these `function`/`class` container scripts into a PowerShell module (see: #18740).

**Comment by rhubarb-geek-nz** [comment]
> Btw. a additional feature of a `Using Script` directive (as apposed to dot-sourcing) could be that it only includes specific definitions as `function`, `class`, `enum` and `filter`

I am not sure what you mean here, a module should only contain definitions of cmdlets, functions etc that are then exported for use outside the module or used internally. I would expect that the only thing that should be invoked from the module itself is [Export-ModuleMember](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/export-modulemember) at the top level in the psm1 file.

**Comment by iRon7** [comment]
What I mean (see also #18740), is that it appears quiet common to load (`.ps1`) scripts into a module file (`.psm1) by dot-sourcing them, see e.g.: https://stackoverflow.com/a/44512990/1701026. In this idea, you presumably just want to load functions (including classes, filters, enums etc.). 
The issue with this approach is that you actually *invoke everything* (including cmdlets as `Set-StrictMode`, preference variables, etc.) which could potentially come with each included `.ps1` script. This might accidently break one of the other scripts -parts of the module- (especially if you work together on a single module). Somehow I think it should be possible to just include functions (including classes, filters, enums etc.) instead of a whole script.

**Comment by rhubarb-geek-nz** [comment]
I have used this technique before in modules and it works well. However ps1 scripts that you dot-source into a psm1 file only contain definitions of functions, classes, aliases etc. They don't start running code as if they have been invoked from the command line.

It is the difference between a C, C++ or C# program with "main" and a class library which just contains reusable code. The idea is the module just loads the reusable code.

In the standard UNIX world you have a hierarchy, shell scripts call compiled programs which use static and shared libraries.

In .NET you have programs compiled with  exe extension which reference libraries with dll extension

Same in powershell, you have scripts that you invoke directly from the command line, and then ps1 files you dot-source which only contain definitions of functions, and finally modules which are loaded on demand ( which in turn can dot-source internal ps1 files).

**Comment by iRon7** [comment]
> They don't start running code as if they have been invoked

Correct, I have rephrased that. Dot-sourcing the .ps1 files in a module file *include* everything. The point is that a function or a class can't break another function (or class) if it just been loaded. But everything around it could potentially conflict with other scripts.

**Comment by rhubarb-geek-nz** [comment]
> The point is that a function or a class can't break another function (or class) if it just been loaded. But everything around it could potentially conflict with other scripts.

Hence the idea of modules. Within a module all functions are private unless you export them. So you just export the functions you want to expose, and keep your implementation details private, leaving less chance to conflict.

**Comment by rhubarb-geek-nz** [comment]
> Set-StrictMode -Version Latest

Good point about the [Set-StrictMode](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/set-strictmode?view=powershell-7.4). 

I suggest you have common standards for your module and say that a particular `StrictMode` applies to the whole module and use it as the start of your psm1 file itself so it is set when all source fiiles are read.
